text/0000-native-class-roadmap.md

```diff
@@ -0,0 +1,265 @@
+delegate pattern without dropping support for Ember versions that are still in
+common use.
+
+We can solve #1 during the Ember v3 lifecycle by changing Ember's code
+internally to treat the existing base classes and native classes as delegates.
+However, we cannot count on being able to remove EmberObject for v4, since
+it's likely that there will still be users of Ember versions that do not support
+delegates. So, the overall roadmap would look like:
+
+- Ember v3: Implement delegates in Ember, allowing users to adopt the new world.
+- Ember v4: Deprecate EmberObject, and prepare the framework for removal.
+- Ember v5: Remove EmberObject.
+
+## How we teach this
+
+Once implemented, we can begin providing examples of creating base-less classes
+for all of Ember's standard constructs. We should update the guides to recommend
+this method, make announcments, and move toward officially deprecating
+EmberObject.
+
+We should also write migration guides that help users through the process of
+converting to native classes. In particular, non-Ember constructs that use
+EmberObject will likely have to change a fair amount. Demonstrating various
+conversions in blog posts and guides will help to show people the path forward,
+and spread knowledge of best practices when using native classes.
+
+## Drawbacks
+
+This approach limits what we can do significantly in framework code (no
+inheritance, can't provide defaults, must provide all behavior external to user
+code). There are many established software patterns other than inheritance for
+providing most behavior, so in practice this shouldn't be an issue.
+
+## Alternatives
+
+- **Gradually rewrite current base classes.** We could attempt to gradually rewrite
+Ember constructs such as Services, Routes, and Components in place. This would
+give us more flexibility for a lot of things, like constructor behavior and
+default method implementations, etc. It would probably take about as long to do
+this as the delegate pattern, and could get very messy along the way, especially
+for removing certain constructs (like Mixins).
+- **Provide new base classes at the same time as the old ones.** We could provide
+new base classes, exposed at different import paths, that users could gradually
+switch to over time. This would give us more flexibility as well, but it would
+also be about as much work and the same adoption timeline as the delegate
+pattern, since Ember would have to support both the old Ember object model and
+the new base classes at the same time until the switch. It would also likely be
+fairly confusing to users, since they would have to make sure they were
+importing the correct classes.
+
+## Unresolved questions
+
+If we go this direction, it means we cannot specify any base class behavior. One
+issue that arises is the fact that dependency injections will not be available
+to the class _during_ object construction. They can be assigned _after_ only.
```

Comment:
Does this result affect initializers/instance-initializers?

Reply:
Initializers and instance-initializers are (as I understand them) essentially pure functions that run at the beginning off app creation/instantiation. They would still receive the app definition/instance and container as such, and wouldn't be directly impacted by this change.

However, and injections which are setup by initializers _would_ be affected. All injections are currently assigned by this mechanism, and assigning after would make them unavailable. 

It's worth noting that only potential solutions mentioned in the RFC which would work for injections done this way would be the `init` hook and passing the injections to the constructor as named parameters. Personally, I am not a fan of the ability to unilaterally inject items onto any type thing (e.g. `store` automatically injected to routes/controllers) and think it would make sense to recommend users manually declare the dependency.

Reply:
Perhaps in v4 we have init() to support this and in Ember v5 we drop it and make everything explicit?

Reply:
I think that would be an option if we decided to go with the other DI options and drop `init` altogether. It would give us a good timeline to clean things up on as well, and falls neatly inline with the proposed timeline for removing EmberObject.